core/nandtool/nandtool.c

```diff
@@ -0,0 +1,245 @@
+	unsigned int page, block, boffs, offs = 0;
+	struct stat stat;
+	char *buff;
+	int fd, err;
+
+	if ((err = open(path, O_RDONLY)) < 0) {
+		fprintf(stderr, "nandtool: failed to open file %s, err: %d\n", path, err);
+		return err;
+	}
+	fd = err;
+
+	if ((buff = malloc(pagesz)) == NULL) {
+		err = -ENOMEM;
+		fprintf(stderr, "nandtool: failed to allocate buffer, err: %d\n", err);
+		close(fd);
+		return err;
+	}
+
+	do {
+		if ((err = fstat(fd, &stat)) < 0) {
+			fprintf(stderr, "nandtool: failed to stat file %s, err: %d\n", path, err);
+			break;
+		}
+
+		start *= npages;
+		while (offs < stat.st_size) {
+			page = start + (offs / pagesz);
+			block = page / npages;
+
+			if ((err = flashmng_isbad(nandtool_common.oid, block)) < 0) {
+				fprintf(stderr, "nandtool: failed to check block %u, err: %d\n", block, err);
+				break;
+			}
+			else if (err > 0) {
+				start++;
+				continue;
+			}
+
+			memset(buff, 0, pagesz);
+			for (boffs = 0; boffs < pagesz; boffs += err) {
+				if ((err = read(fd, buff + boffs, pagesz - boffs)) <= 0) {
+					if (err) {
+						err = -errno;
+						fprintf(stderr, "nandtool: failed to read file %s, err: %d\n", path, err);
+					}
+					break;
+				}
+			}
+
+			if (err < 0)
+				break;
+
+			if (raw) {
+				if ((err = flashmng_writeraw(nandtool_common.oid, page, buff, pagesz)) < 0) {
+					fprintf(stderr, "nandtool: failed to write raw data to page %u, err: %d\n", page, err);
+					break;
+				}
+			}
+			else {
+				if ((err = lseek(nandtool_common.fd, (off_t)page * pagesz, SEEK_SET)) < 0) {
+					fprintf(stderr, "nandtool: failed to lseek to page %u, err: %d\n", page, err);
+					break;
+				}
+
+				if ((err = write(nandtool_common.fd, buff, pagesz)) != pagesz) {
+					err = -errno;
+					fprintf(stderr, "nandtool: failed to write data to page %u, err: %d\n", page, err);
+					break;
+				}
+			}
+
+			offs += boffs;
+			printf("\rFlashing %s %u%%...", path, (unsigned int)(100 * offs / stat.st_size));
+		}
+
+		if (err < 0)
+			break;
+
+		printf("\rFlashing %s completed!\n", path);
+	} while (0);
+
+	free(buff);
+	close(fd);
+
+	return err;
+}
+
+
+static int nandtool_erase(unsigned int start, unsigned int end)
+{
+	int err;
+
+	if (end < start) {
+		fprintf(stderr, "nandtool: invalid erase range <%d:%d)\n", start, end);
+		return -EINVAL;
+	}
+
+	if ((err = flashmng_erase(nandtool_common.oid, start, end)) < 0)
+		fprintf(stderr, "nandtool: failed to erase blocks, err: %d\n", err);
+
+	return EOK;
+}
+
+
+static int nandtool_check(void)
+{
+	int err;
+
+	if ((err = flashmng_checkbad(nandtool_common.oid)) < 0)
+		fprintf(stderr, "nandtool: failed to complete bad block check, err: %d\n", err);
+
+	return EOK;
+}
+
+
+static void nandtool_help(const char *prog)
+{
+	printf("Usage: %s [options] <device>\n", prog);
+	printf("\t-e <start[:end]> - erase block(s) in <start:end) range\n");
+	printf("\t-i <path>        - file path (requires -s option)\n");
+	printf("\t-r <path>        - just like -i option but raw\n");
```

Comment:
not sure why I need to put file path separately for normal and raw cases (`raw` should not have filepath param?)

Reply:
Changed as suggested, we pass the file to be flashed with `-i` option and `-r` acts as a flag for flashing raw data.